### PR TITLE
gk: change prefetches from batched to interleaved

### PR DESCRIPTION
**Pull request by mengxiang0811**
After applying this patch, below is the performance:

| lcores | table size | GK Mpps rcvd (0%) | GK Mpps rcvd (50%) | GK Mpps rcvd (99%) | GK Mpps rcvd (mean) | cli Mpps sent (0%) | cli Mpps sent (50%) | cli Mpps sent (99%) | cli Mpps sent (mean) |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|
|  1  |  2^15  |       2.54        |        2.86        |        3.18        |        2.86         |       11.09        |        11.62        |        12.13        |         11.62         | 
|  1  |  2^20  |       1.92        |        2.45        |        2.59        |        2.38         |       9.73        |        11.56        |        12.52        |         11.48         | 
|  1  |  2^22  |       1.61        |        2.49        |        2.64        |        2.33         |       10.99        |        11.65        |        12.26        |         11.71         | 
|        |            |                   |                    |                    |                     |                    |                     |                     |                      |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|
|  2  |  2^15  |       0.94        |        2.37        |        2.54        |        2.1         |       9.78        |        11.48        |        11.82        |         11.29         | 
|  2  |  2^20  |       1.43        |        2.12        |        2.19        |        1.99         |       10.32        |        11.36        |        11.71        |         11.39         | 
|  2  |  2^22  |       0.9        |        1.71        |        1.82        |        1.6         |       10.23        |        10.72        |        11.08        |         10.72         | 
|        |            |                   |                    |                    |                     |                    |                     |                     |                      |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|
|  3  |  2^15  |       1.7        |        2.76        |        3.97        |        2.69         |       10.13        |        11.06        |        11.33        |         11.0         | 
|  3  |  2^20  |       1.16        |        2.53        |        2.81        |        2.28         |       9.54        |        10.84        |        11.49        |         10.84         | 
|  3  |  2^22  |       0.95        |        1.95        |        3.39        |        2.03         |       9.7        |        10.83        |        11.93        |         10.89         | 
|        |            |                   |                    |                    |                     |                    |                     |                     |                      |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|

### CODE REVIEW

File: gk/main.c [hunk 1]
```diff
@@ -1727,11 +1729,51 @@ process_pkts_front(uint16_t port_front, uint16_t port_back,
...
+		rte_prefetch0(rte_pktmbuf_mtod_offset(rx_bufs[i], void *, 0));
 
 	/* Extract packet and flow information. */
-	for (i = 0; i < num_rx; i++) {
+	for (i = 0; i < (num_rx - PREFETCH_OFFSET); i++) {
+		struct ipacket *packet = &pkt_arr[num_ip_flows];
+		struct rte_mbuf *pkt = rx_bufs[i];
+
+		rte_prefetch0(rte_pktmbuf_mtod_offset(
+			rx_bufs[i + PREFETCH_OFFSET], void *, 0));
```
Comment (AltraMayor) [inline gk/main.c]: The only difference between the body of this `for` loop and the following `for` loop is the prefetch above. Write a function for the body of this `for` loop without the prefetch and avoid the duplicate code. If measurements point that having a function here presents a performance issue, make that function inline.

**Comment by AltraMayor** [comment]
About the evaluation, it is already impossible for me to figure out which other evaluation I should use as a reference to interpret the numbers. Let's minimize links between pull requests because each pull request follows a different path toward merge. Thus, add the evaluation of the master branch to this pull request.

Given that we are dealing with subtle prefetch improvements here, add the perf report on loads for before and after this pull request.

File: gk/main.c (hunk 1 above)
Comment (mengxiang0811) [reply gk/main.c]: After evaluating the updated patch, there is no performance issue.

**Comment by mengxiang0811** [comment]
Below is the performance of the master branch:


| lcores | table size | GK Mpps rcvd (0%) | GK Mpps rcvd (50%) | GK Mpps rcvd (99%) | GK Mpps rcvd (mean) | cli Mpps sent (0%) | cli Mpps sent (50%) | cli Mpps sent (99%) | cli Mpps sent (mean) |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|
|  1  |  2^15  |       1.93        |        2.81        |        2.97        |        2.53         |       10.13        |        11.4        |        11.99        |         11.26         | 
|  1  |  2^20  |       0.72        |        1.29        |        1.63        |        1.28         |       9.02        |        11.12        |        11.78        |         10.77         | 
|  1  |  2^22  |       1.35        |        1.65        |        2.19        |        1.8         |       10.48        |        11.38        |        11.69        |         11.24         | 
|        |            |                   |                    |                    |                     |                    |                     |                     |                      |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|
|  2  |  2^15  |       0.77        |        1.46        |        1.5        |        1.28         |       9.62        |        10.34        |        10.96        |         10.27         | 
|  2  |  2^20  |       1.07        |        1.57        |        1.65        |        1.5         |       9.69        |        10.94        |        11.78        |         10.99         | 
|  2  |  2^22  |       1.21        |        1.57        |        1.73        |        1.54         |       10.01        |        10.65        |        11.28        |         10.55         | 
|        |            |                   |                    |                    |                     |                    |                     |                     |                      |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|
|  3  |  2^15  |       2.24        |        2.63        |        2.77        |        2.58         |       9.89        |        10.44        |        11.04        |         10.45         | 
|  3  |  2^20  |       1.24        |        2.51        |        2.71        |        2.31         |       9.77        |        10.47        |        10.99        |         10.38         | 
|  3  |  2^22  |       0.46        |        1.74        |        2.42        |        1.62         |       9.34        |        10.61        |        11.28        |         10.63         | 
|        |            |                   |                    |                    |                     |                    |                     |                     |                      |
|--------|------------|-------------------|--------------------|--------------------|---------------------|--------------------|---------------------|---------------------|----------------------|

**Comment by mengxiang0811** [comment]
In the tests, we use 3 lcores and 2^22-entry flow tables.

(1) **loads of master branch**

```
Overhead  Command        Shared Object      Symbol                                                                                                                               
   2.63%  lcore-slave-1  gatekeeper         [.] rte_hash_cuckoo_make_space_mw                                                                                                    ▒
   1.87%  lcore-slave-1  gatekeeper         [.] ixgbe_recv_pkts_vec                                                                                                              ▒
   1.73%  lcore-slave-1  gatekeeper         [.] __rte_hash_lookup_bulk_with_hash_l.constprop.29                                                                                  ▒
   1.27%  lcore-slave-1  [kernel.kallsyms]  [k] skb_release_data                                                                                                                 ▒
   1.13%  lcore-slave-1  [kernel.kallsyms]  [k] skb_release_head_state                                                                                                           ▒
   1.01%  lcore-slave-1  [kernel.kallsyms]  [k] sock_def_write_space                                                                                                             ▒
   0.93%  lcore-slave-1  [kernel.kallsyms]  [k] sock_wfree                                                                                                                       ▒
   0.89%  lcore-slave-1  gatekeeper         [.] encapsulate                                                                                                                      ▒
   0.88%  lcore-slave-1  gatekeeper         [.] process_pkts_front.constprop.20                                                                                                  ▒
   0.87%  lcore-slave-1  gatekeeper         [.] gk_process_request.isra.16                                                                                                       ▒
   0.75%  lcore-slave-1  gatekeeper         [.] common_ring_mc_dequeue                                                                                                           ▒
   0.69%  lcore-slave-1  [kernel.kallsyms]  [k] napi_consume_skb                                                                                                                 ▒
   0.67%  lcore-slave-1  gatekeeper         [.] adjust_pkt_len                                                                                                                   ▒
   0.65%  lcore-slave-1  [kernel.kallsyms]  [k] __indirect_thunk_start                                                                                                           ▒
   0.49%  lcore-slave-1  [kernel.kallsyms]  [k] cmpxchg_double_slab.isra.61                                                                                                      ▒
   0.39%  lcore-slave-1  [kernel.kallsyms]  [k] ixgbe_poll                                                                                                                       ▒
   0.30%  lcore-slave-1  gatekeeper         [.] lookup_fib_bulk                                                                                                                  ▒
   0.26%  lcore-slave-1  [kernel.kallsyms]  [k] __slab_free                                                                                                                      ▒
   0.21%  lcore-slave-1  gatekeeper         [.] gk_solicitor_enqueue_bulk                                                                                                        ▒
   0.14%  lcore-slave-1  gatekeeper         [.] __rte_hash_del_key_with_hash                                                                                                     ▒
   0.13%  lcore-slave-1  gatekeeper         [.] ixgbe_xmit_pkts                                                                                                                  ▒
   0.11%  lcore-slave-1  gatekeeper         [.] extract_packet_info                                                                                                              ▒
   0.11%  lcore-slave-1  gatekeeper         [.] rte_hash_lookup_with_hash                                                                                                        ▒
   0.10%  lcore-slave-1  [kernel.kallsyms]  [k] kfree                                                                                                                            ▒
   0.10%  lcore-slave-1  gatekeeper         [.] gk_proc                                                                                                                          ▒
   0.09%  lcore-slave-1  gatekeeper         [.] process_cmds_from_mailbox                                                                                                        ▒
   0.08%  lcore-slave-1  [kernel.kallsyms]  [k] kmem_cache_free_bulk                                                                                                             ▒
   0.08%  lcore-slave-1  gatekeeper         [.] __rte_hash_add_key_with_hash                                                                                                     ▒
   0.07%  lcore-slave-1  gatekeeper         [.] common_ring_mp_enqueue                                                                                                           ▒
   0.06%  lcore-slave-1  gatekeeper         [.] ixgbe_recv_scattered_pkts_vec                                                                                                    ▒
   0.06%  lcore-slave-1  [kernel.kallsyms]  [k] dql_completed                                                                                                                    ▒
   0.05%  lcore-slave-1  [kernel.kallsyms]  [k] skb_release_all                                                                                                                  ▒
   0.04%  lcore-slave-1  libc-2.27.so       [.] __memmove_sse2_unaligned_erms                                                                                                    ▒
   0.04%  lcore-slave-1  [kernel.kallsyms]  [k] __free_pages_ok                                                                                                                  ▒
   0.04%  lcore-slave-1  gatekeeper         [.] memcmp@plt                                                                                                                       ▒
   0.03%  lcore-slave-1  gatekeeper         [.] rss_ip_flow_hf                                                                                                                   ▒
   0.03%  lcore-slave-1  [kernel.kallsyms]  [k] unfreeze_partials.isra.71                                                                                                        ▒
   0.03%  lcore-slave-1  [kernel.kallsyms]  [k] ixgbe_clean_rx_irq                                                                                                               ▒
   0.03%  lcore-slave-1  gatekeeper         [.] rte_hash_bucket_iterate                                                                                                          ▒
   0.03%  lcore-slave-1  libc-2.27.so       [.] __memcmp_sse4_1                                                                                                                  ▒
   0.02%  lcore-slave-1  [kernel.kallsyms]  [k] put_cpu_partial                                                                                                                  ▒
   0.02%  lcore-slave-1  gatekeeper         [.] gk_bpf_decide_pkt 
```

(2) **loads after applying this patch**

```
Overhead  Command        Shared Object      Symbol                                                                                                                               ◆
   4.02%  lcore-slave-1  gatekeeper         [.] rte_hash_cuckoo_make_space_mw                                                                                                    ▒
   2.28%  lcore-slave-1  gatekeeper         [.] ixgbe_recv_pkts_vec                                                                                                              ▒
   1.94%  lcore-slave-1  gatekeeper         [.] __rte_hash_lookup_bulk_with_hash_l.constprop.29                                                                                  ▒
   1.36%  lcore-slave-1  gatekeeper         [.] encapsulate                                                                                                                      ▒
   1.33%  lcore-slave-1  gatekeeper         [.] extract_packet_info                                                                                                              ▒
   1.13%  lcore-slave-1  gatekeeper         [.] common_ring_mc_dequeue                                                                                                           ▒
   0.93%  lcore-slave-1  gatekeeper         [.] gk_process_request.isra.16                                                                                                       ▒
   0.72%  lcore-slave-1  gatekeeper         [.] adjust_pkt_len                                                                                                                   ▒
   0.34%  lcore-slave-1  gatekeeper         [.] lookup_fib_bulk                                                                                                                  ▒
   0.34%  lcore-slave-1  gatekeeper         [.] process_pkts_front.constprop.20                                                                                                  ▒
   0.22%  lcore-slave-1  gatekeeper         [.] gk_solicitor_enqueue_bulk                                                                                                        ▒
   0.18%  lcore-slave-1  gatekeeper         [.] rte_hash_lookup_with_hash                                                                                                        ▒
   0.12%  lcore-slave-1  gatekeeper         [.] ixgbe_xmit_pkts                                                                                                                  ▒
   0.12%  lcore-slave-1  gatekeeper         [.] common_ring_mp_enqueue                                                                                                           ▒
   0.12%  lcore-slave-1  gatekeeper         [.] __rte_hash_del_key_with_hash                                                                                                     ▒
   0.08%  lcore-slave-1  gatekeeper         [.] __rte_hash_add_key_with_hash                                                                                                     ▒
   0.07%  lcore-slave-1  gatekeeper         [.] process_cmds_from_mailbox                                                                                                        ▒
   0.07%  lcore-slave-1  gatekeeper         [.] gk_proc                                                                                                                          ▒
   0.07%  lcore-slave-1  gatekeeper         [.] memcmp@plt                                                                                                                       ▒
   0.06%  lcore-slave-1  gatekeeper         [.] ixgbe_recv_scattered_pkts_vec                                                                                                    ▒
   0.04%  lcore-slave-1  libc-2.27.so       [.] __memmove_sse2_unaligned_erms                                                                                                    ▒
   0.03%  lcore-slave-1  libc-2.27.so       [.] __memcmp_sse4_1                                                                                                                  ▒
   0.03%  lcore-slave-1  gatekeeper         [.] rte_hash_bucket_iterate                                                                                                          ▒
   0.03%  lcore-slave-1  gatekeeper         [.] gk_bpf_decide_pkt                                                                                                                ▒
   0.02%  lcore-slave-1  gatekeeper         [.] rss_ip_flow_hf                                                                                                                   ▒
   0.01%  lcore-slave-1  gatekeeper         [.] ip_flow_cmp_eq                                                                                                                   ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] __indirect_thunk_start                                                                                                           ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] perf_event_task_tick                                                                                                             ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] ixgbe_poll                                                                                                                       ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] rcu_check_callbacks                                                                                                              ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] update_load_avg                                                                                                                  ▒
   0.00%  lcore-slave-1  gatekeeper         [.] rte_hash_lookup_bulk_with_hash                                                                                                   ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] skb_release_head_state                                                                                                           ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] update_min_vruntime                                                                                                              ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] napi_consume_skb                                                                                                                 ▒
   0.00%  lcore-slave-1  gatekeeper         [.] enqueue_slot_back.isra.11.part.12                                                                                                ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] skb_release_data                                                                                                                 ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] sock_def_write_space                                                                                                             ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] sock_wfree                                                                                                                       ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] __acct_update_integrals                                                                                                          ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] hrtimer_active                                                                                                                   ▒
   0.00%  lcore-slave-1  [kernel.kallsyms]  [k] update_curr
```

Note that although the loads of ``process_pkts_front()`` has decreased, the loads of ``extract_packet_info()``.

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.

**Pull request merged**